bindinfo/cache.go

```diff
@@ -0,0 +1,191 @@
+	bc := h.Load()
+	if bc != nil {
+		return bc.(map[string][]*bindMeta)
+	}
+	return make(map[string][]*bindMeta)
+}
+
+// LoadDiff is used to load new bind info to cache bc.
+func (bindCacheUpdater *BindCacheUpdater) loadDiff(sql string, bc cache) error {
+	recordSets, err := bindCacheUpdater.ctx.(sqlexec.SQLExecutor).Execute(context.Background(), sql)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	rs := recordSets[0]
+	defer terror.Call(rs.Close)
+	chkBatch := rs.NewRecordBatch()
+	for {
+		err = rs.Next(context.TODO(), chkBatch)
+		if err != nil || chkBatch.NumRows() == 0 {
+			return errors.Trace(err)
+		}
+
+		it := chunk.NewIterator4Chunk(chkBatch.Chunk)
+		for row := it.Begin(); row != it.End(); row = it.Next() {
+			record := newBindMeta(row)
+			err = bc.appendNode(record, bindCacheUpdater.parser)
+			if err != nil {
+				return err
+			}
+			if record.UpdateTime.Compare(bindCacheUpdater.lastUpdateTime) == 1 {
+				bindCacheUpdater.lastUpdateTime = record.UpdateTime
+			}
+		}
+	}
+}
+
+// Update updates the BindCacheUpdater's cache. The fullLoad is true when tidb first startup, otherwise it is false.
+func (bindCacheUpdater *BindCacheUpdater) Update(fullLoad bool) (err error) {
+	var sql string
+	bc := bindCacheUpdater.globalHandle.Get()
+	newBc := make(map[string][]*bindMeta, len(bc))
+	for hash, bindDataArr := range bc {
+		newBc[hash] = append(newBc[hash], bindDataArr...)
+	}
+
+	if fullLoad {
+		sql = "select original_sql, bind_sql, default_db, status, create_time, update_time, charset, collation from mysql.bind_info"
+	} else {
+		sql = fmt.Sprintf("select original_sql, bind_sql, default_db, status, create_time, update_time, charset, collation from mysql.bind_info where update_time > \"%s\"", bindCacheUpdater.lastUpdateTime.String())
+	}
+	err = bindCacheUpdater.loadDiff(sql, newBc)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	bindCacheUpdater.globalHandle.Store(newBc)
+	return nil
+}
+
+func newBindMeta(row chunk.Row) *bindRecord {
+	return &bindRecord{
+		OriginalSQL: row.GetString(0),
+		BindSQL:     row.GetString(1),
+		Db:          row.GetString(2),
+		Status:      row.GetString(3),
+		CreateTime:  row.GetTime(4),
+		UpdateTime:  row.GetTime(5),
+		Charset:     row.GetString(6),
+		Collation:   row.GetString(7),
+	}
+}
+
+func (b cache) appendNode(newBindRecord *bindRecord, sparser *parser.Parser) error {
+	hash := parser.DigestHash(newBindRecord.OriginalSQL)
+	if bindArr, ok := b[hash]; ok {
+		for idx, v := range bindArr {
+			if v.OriginalSQL == newBindRecord.OriginalSQL && v.Db == newBindRecord.Db {
+				b[hash] = append(b[hash][:idx], b[hash][idx+1:]...)
+				if len(b[hash]) == 0 {
+					delete(b, hash)
+				}
+				break
+			}
+		}
+	}
+	if newBindRecord.Status == deleted {
+		return nil
+	}
+	stmtNodes, _, err := sparser.Parse(newBindRecord.BindSQL, newBindRecord.Charset, newBindRecord.Collation)
+	if err != nil {
+		log.Warnf("parse error:\n%v\n%s", err, newBindRecord.BindSQL)
```

Comment:
Why do we need to print a log here?

Reply:
return the error and stop the updating operation?